src/lib/sparse_ledger_lib/sparse_ledger.ml

```diff
@@ -173,6 +184,80 @@ end = struct
     ; indexes = (account_id, index) :: t.indexes
     }
 
+  let add_wide_path_unsafe tree0 path0 account =
```

Comment:
I would propose that we stick with the original control flow because it's easier to understand plus the original implementation has some assertion to check the correctness of the merkle path. If you want to skip the assertions, you could also easily do that:
```
  let add_wide_path depth0 tree0 path0 account =
    let rec build_tree h height p =
      match p with
      | `Left (h_l, h_r) :: path ->
        let l = build_tree h_l (height - 1) path in 
        Tree.Node (h, l, Hash h_r)
      | `Right (h_l, h_r) :: path ->
        let r = build_tree h_r (height - 1) path in 
        Node (h, Hash h_l, r)
      | [] ->
        assert (height = -1) ;
        Account account 
      in
      let rec union height tree path =
        match (tree, path) with 
        | Tree.Hash h, path ->
          let t = build_tree h height path in
          [%test_result: Hash.t]
            ~message:
              "Hashes in union are not equal, something is wrong with your \
               ledger"
            ~expect:h (hash t) ;
          t
        | Node (h, l, r), `Left _ :: path ->
          let l = union (height - 1) l path in 
          Node (h, l, r)
        | Node (h, l, r), `Right _ :: path ->
          let r = union (height - 1) r path in 
          Node (h, l, r) 
        | Node _, [] ->
          failwith "Path too short" 
        | Account _, _ :: _ ->
          failwith "Path too long" 
        | Account a, [] ->
          assert (Account.equal a account) ;
          tree 
        in 
        union (depth0 - 1) tree0 (List.rev path0)  
```